template.tpl

```diff
@@ -0,0 +1,1575 @@
+      const setVal = interpret(m, 'default', 'eventProperty', evObj);
+      setFromPath(m, setVal, selfDescData);
+    } else {
+      const setVal = interpret(m.value, m.type, m.ref, evObj);
+      setFromPath(m.spPropName, setVal, selfDescData);
+    }
+  });
+
+  const uePr = JSON.stringify({
+    schema: spSelfDescribingSchema,
+    data: {
+      schema: evSchema,
+      data: selfDescData
+    }
+  });
+
+  if (encodeBase64) {
+    event.ue_px = base64urlencode(uePr);
+  } else {
+    event.ue_pr = uePr;
+  }
+
+  return event;
+};
+
+/*
+ * Given a client event, makes a structured Snowplow event
+ *
+ * @param evObj {Object} - the client event object
+ * @returns - the structured Snowplow event object
+ */
+const mkStructuredEvent = (evObj) => {
+  const action = evObj.event_action ? evObj.event_action : evObj.event_name;
+  const category = evObj.event_category;
+
+  if (action && category) {
+    const event = mkStandardPairs(evObj, 'se');
+    event.se_ac = action;
+    event.se_ca = category;
+    event.se_la = evObj.event_label;
+    event.se_va = evObj.event_value;
+
+    return event;
+  }
+  return undefined;
+};
+
+/*
+ * Makes a Snowplow event from a non-Snowplow client event.
+ *
+ * @param evObj {Object} - the client event object
+ * @param customDefs {Object} - the custom event data definitions
+ * @param encode {boolean} - whether base64 encoding will be used
+ * @returns - the Snowplow event or undefined, if event could not be constructed
+ */
+const mkSnowplowEvent = (evObj, customDefs, encode) => {
+  const eventName = evObj.event_name;
+
+  if (!customDefs.hasOwnProperty(eventName)) {
+    return undefined;
+  }
+
+  if (customDefs[eventName].eventType === 'ue') {
+    const eventSchema = asIgluSchema(customDefs[eventName].eventSchema);
+    const eventMappings = customDefs[eventName].eventDataMappings;
+
+    return mkSelfDescribingEvent(evObj, eventSchema, eventMappings, encode);
+  }
+
+  if (customDefs[eventName].eventType === 'se') {
+    return mkStructuredEvent(evObj);
+  }
+
+  return undefined;
+};
+
+/*
+ * Given the client event object,
+ *  - gets the Snowplow event from 'x-sp-tp2' (by Snowplow client)
+ *  - or makes a Snowplow event
+ */
+const buildSpEvent = (evObj, customDefs, encode) => {
+  if (evObj["x-sp-tp2"]) {
+    const spEvent = evObj["x-sp-tp2"];
+    spEvent.ip = evObj.ip_override;
+
+    return spEvent;
+  }
+  const rawEvent = mkSnowplowEvent(evObj, customDefs, encode);
+
+  return rawEvent ? rawEvent : undefined;
+};
+
+/*
+ * Given a Snowplow event, returns the Snowplow payload to be sent.
+ */
+const mkSnowplowPayload = (snowplowEvent) => {
+  if (snowplowEvent) {
+    const snowplowPayload = {
+      schema: spPayloadDataSchema,
+      data: [ snowplowEvent ]
+    };
+    return snowplowPayload;
+  }
+  return undefined;
+};
+
+// Main
+if (!data.collectorUrl) {
+  return fail('A collector URL must be specified');
+}
+
+const spCustomDefs = mkCustomDefs(data, spDefaultCustomDefs);
```

Comment:
We're doing this work even if it's a `x-sp-tp2` event but as far as I can see, we only use these custom definitions for other events. `data` and `spDefaultCustomDefs` are globals, so I think we can move this inside `buildSpEvent`?

Reply:
Changing now! Indeed, no need for these to run for `x-sp-tp2` events.